File: android/app/src/main/kotlin/dev/omnesis/android/ui/phonesetup/NotificationsSetupStep.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.phonesetup

import android.Manifest
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.AccessTime
import androidx.compose.material.icons.outlined.NotificationsActive
import androidx.compose.material.icons.outlined.Shield
import androidx.compose.material.icons.outlined.SyncProblem
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import dev.omnesis.android.notifications.FcmPushManager
import dev.omnesis.android.session.SessionManager
import dev.omnesis.android.setup.PhoneSetupStep
import dev.omnesis.android.setup.SetupHighlight
import dev.omnesis.android.setup.SetupStepController
import dev.omnesis.android.setup.SetupStepCopy
import dev.omnesis.android.setup.flow.NotificationPermissionState
import dev.omnesis.android.setup.flow.SetupAvailability
import dev.omnesis.android.setup.flow.SetupGroup
import dev.omnesis.android.setup.flow.SetupOutcome
import dev.omnesis.android.setup.flow.SourceEnableActions
import dev.omnesis.android.setup.flow.SourceEnableSequence
import dev.omnesis.android.setup.flow.notificationRow
import dev.omnesis.android.setup.ui.rememberPermissionLauncher
import dev.omnesis.android.setup.ui.rememberSourceSetupController
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

val NotificationsSetupCopy = SetupStepCopy(
    name = "Notifications",
    glyph = Icons.Outlined.NotificationsActive,
    tint = Color(0xFFFFD60A),
    row = "Answers and approvals",
    value = "Know when a slow answer is ready, when something needs your approval, and when a source stops syncing.",
    highlights = listOf(
        SetupHighlight(Icons.Outlined.AccessTime, "Answers that took a while"),
        SetupHighlight(Icons.Outlined.Shield, "Requests waiting for your approval"),
        SetupHighlight(Icons.Outlined.SyncProblem, "Sources that need attention"),
    ),
    fine = "Android asks next. You can change this in Settings any time.",
    permissionLabel = "notifications",
    onTitle = "Notifications are on",
    onBody = "You'll hear when an answer is ready, when something needs your approval, or when a source needs attention.",
    offTitle = "Notifications are off",
    offBody = "You can turn them on in Settings any time.",
    settingsSteps = "In Settings, turn on notifications for Omnesis.",
    workingLabel = "Turning on notifications…",
)

/** How a resolved Notifications page reads once the grant is known again; a declined page stays declined. */
internal fun notificationsRefreshedOutcome(previous: SetupOutcome, granted: Boolean): SetupOutcome = when {
    previous == SetupOutcome.Skipped -> previous
    granted -> SetupOutcome.On
    else -> SetupOutcome.NotAllowed
}

/**
 * The Notifications page of the phone setup flow: the one place, beside the
 * button in Settings, that asks Android for the notification permission.
 * Offered only by a build that can deliver notifications.
 */
@Singleton
class NotificationsSetupStep @Inject constructor(
    private val pushManager: FcmPushManager,
    private val session: SessionManager,
) : PhoneSetupStep {
    /** Outlives any page, so Android's answer is kept however the page that asked went away. */
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    internal val sequence = SourceEnableSequence(
        scope = scope,
        actions = SourceEnableActions.Local,
        isOn = ::isOn,
        isGranted = { isOn() },
        outcomeAfterEnable = { if (isOn()) SetupOutcome.On else SetupOutcome.NotAllowed },
    )

    override val id: String = "notifications"
    override val order: Int = 100
    override val group: SetupGroup = SetupGroup.ALSO
    override val sourceId: String? = null
    override val copy = NotificationsSetupCopy
    override val outcomes: Flow<SetupOutcome?> = sequence.outcomes

    override fun availability(): SetupAvailability =
        notificationRow(id, pushManager.notificationPermissionState(), pushManager.configured).availability

    override fun isOn(): Boolean = pushManager.notificationPermissionState() == NotificationPermissionState.GRANTED

    override suspend fun refreshOutcome(previous: SetupOutcome): SetupOutcome? = notificationsRefreshedOutcome(previous, isOn())

    override fun notNow() = sequence.notNow()

    /** Back on this page while it is unresolved: the gateway hears where delivery stands, and a grant made in Settings is read. */
    override fun resume() {
        reportAnswer()
        sequence.onResume()
    }

    override fun reset() = sequence.reset()

    @Composable
    override fun rememberController(): SetupStepController {
        val launcher = rememberPermissionLauncher(
            permissions = {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) arrayOf(Manifest.permission.POST_NOTIFICATIONS) else emptyArray()
            },
            rationalePermission = null,
            accessGranted = { grants -> grants.isNotEmpty() && grants.values.all { it } },
            onAnswer = { granted, _ ->
                pushManager.markNotificationPermissionPrompted()
                reportAnswer()
                sequence.onAccessResult(granted)
            },
            settingsIntent = ::notificationSettingsIntent,
        )
        return rememberSourceSetupController(sequence = sequence, launcher = launcher)
    }

    /**
     * Tells the gateway where notification delivery stands and registers for
     * push at once, now that Android answered: a relay request the gateway
     * needs then usually arrives while setup is still showing, which asks on
     * its own page before Finish.
     */
    private fun reportAnswer() {
        scope.launch {
            try {
                session.reportNotificationHealth()
                session.retryFcmRegistration()
            } catch (e: CancellationException) {
                throw e
            } catch (_: Exception) {
                // Delivery health is reported again on the next foreground.
            }
        }
    }
}

private fun notificationSettingsIntent(context: Context): Intent =
    Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS).putExtra(Settings.EXTRA_APP_PACKAGE, context.packageName)

File: ios/Sources/Omnesis/Notifications/NotificationsSetupStep.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation

/// The user's notification decision, as setup reads it.
public enum PhoneSetupNotificationPermission: Equatable, Sendable {
    case notDetermined
    case authorized
    case denied
}

/// What the Notifications setup step needs from the app.
@MainActor
public protocol NotificationsSetupHost: AnyObject {
    var notificationPermission: PhoneSetupNotificationPermission { get }
    /// A relay-consent decision waiting for the owner, if the gateway needs one.
    var relayPushConsentRequest: RelayPushConsentRequest? { get }
    func refreshPushDeliveryHealth() async
    /// Shows the iOS notification prompt and registers for pushes when
    /// allowed. Returns whether notifications are allowed afterwards.
    func requestNotificationPermission() async -> Bool
    func allowRelayPush(_ request: RelayPushConsentRequest) async throws
    func dismissRelayPushConsent(_ request: RelayPushConsentRequest)
}

/// Notifications' page in phone setup. It sends nothing to the gateway, so it
/// lists what it alerts about instead of a ledger.
public struct NotificationsSetupStep: PhoneSetupStep {
    public static let stepId = "notifications"

    public static let copy = PhoneSetupCopy(
        title: "Notifications",
        row: "Answers and approvals",
        value: "Know when a slow answer is ready, when something needs your approval, and when a source stops syncing.",
        highlights: [
            PhoneSetupHighlight(symbol: "hourglass", text: "Answers that took a while"),
            PhoneSetupHighlight(symbol: "checkmark.seal", text: "Requests waiting for your approval"),
            PhoneSetupHighlight(symbol: "exclamationmark.triangle", text: "Sources that need attention"),
        ],
        fine: "iOS asks next. You can change this in Settings any time.",
        permissionLabel: "notifications",
        onBody: "You'll hear when an answer is ready, when something needs your approval, or when a source needs attention.",
        onTitle: "Notifications are on",
        offTitle: "Notifications are off",
        offBody: "You can turn them on in Settings any time.",
        notAllowedSettingsSteps: "In Settings, tap Notifications and turn on Allow Notifications.",
        busyLabel: "Turning on notifications…",
        tint: 0xFFD60A,
        symbol: "bell.fill"
    )

    static let offInSettingsReason = "Off in Settings"

    unowned let host: any NotificationsSetupHost

    public var id: String {
        Self.stepId
    }

    public var group: PhoneSetupStepGroup {
        .also
    }

    public var copy: PhoneSetupCopy {
        Self.copy
    }

    public var rowState: PhoneSetupRowState {
        switch host.notificationPermission {
        case .notDetermined: .selectable
        case .authorized: .alreadyOn
        case .denied: .unavailable(reason: Self.offInSettingsReason)
        }
    }

    public var authorization: MobileSourceAuthorization? {
        switch host.notificationPermission {
        case .authorized: .granted(.full)
        case .denied: .notAllowed
        case .notDetermined: nil
        }
    }

    public func refresh() async {
        await host.refreshPushDeliveryHealth()
    }

    public func enable(choice _: MobileSourceActivationChoice?) async -> PhoneSetupOutcome {
        await host.requestNotificationPermission() ? .on : .notAllowed
    }

    public func currentOutcome() -> PhoneSetupOutcome? {
        switch host.notificationPermission {
        case .notDetermined: nil
        case .authorized: .on
        case .denied: .notAllowed
        }
    }
}

extension PushDeliveryHealth {
    /// The permission decision behind this delivery state.
    public var setupPermission: PhoneSetupNotificationPermission {
        switch self {
        case .notDetermined: .notDetermined
        case .permissionDenied: .denied
        case .ok, .scheduledSummary, .alertsOff: .authorized
        }
    }

    /// Whether the app registers for pushes without asking. Omnesis only
    /// shows the iOS notification prompt when the user asks for
    /// notifications, so an undecided permission registers nothing.
    public var registersPushTokenAutomatically: Bool {
        self != .notDetermined
    }

    /// Whether the state earns a warning banner on home. An undecided
    /// permission is the user's choice to make in setup or Settings, not a
    /// fault to flag.
    public var raisesHomeAttention: Bool {
        warning != nil && self != .notDetermined
    }
}
